// linaria/lib/rollup.js
"use strict";

function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

var _require = require('rollup-pluginutils'),
    createFilter = _require.createFilter;

var _transform = require('./transform');

var slugify = require('./slugify');

module.exports = function linaria() {
  var _ref = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {},
      include = _ref.include,
      exclude = _ref.exclude,
      sourceMap = _ref.sourceMap,
      rest = _objectWithoutProperties(_ref, ["include", "exclude", "sourceMap"]);

  var filter = createFilter(include, exclude);
  var cssLookup = {};
  return {
    name: 'linaria',
    load: function load(id) {
      return cssLookup[id];
    },

    /* eslint-disable-next-line consistent-return */
    resolveId: function resolveId(importee) {
      if (importee in cssLookup) return importee;
    },
    transform: function transform(code, id) {
      if (!filter(id)) return;

      var result = _transform(code, {
        filename: id,
        pluginOptions: rest
      });

      if (!result.cssText) return;
      var cssText = result.cssText;
      var slug = slugify(id);
      var filename = "".concat(id.replace(/\.js$/, ''), "_").concat(slug, ".css");

      if (sourceMap && result.cssSourceMapText) {
        var map = Buffer.from(result.cssSourceMapText).toString('base64');
        cssText += "/*# sourceMappingURL=data:application/json;base64,".concat(map, "*/");
      }

      cssLookup[filename] = cssText;
      result.code += "\nimport ".concat(JSON.stringify(filename), ";\n");
      /* eslint-disable-next-line consistent-return */

      return {
        code: result.code,
        map: result.sourceMap
      };
    }
  };
};
